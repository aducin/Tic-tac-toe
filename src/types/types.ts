interface BoxBase {
    icon: string;
    iconWinning: string;
    value: string;
};

export interface Box extends BoxBase {
    assigned: boolean;
    winning: boolean;
};

export type BoxesProps = {
    boxes: Box[];
    finished: boolean;
    onBoxClicked: (id: number) => void;
    started: boolean;
};

export type BoxItemProps = {
    data: {
        id: number;
        item: Box,
    };
    onBoxClicked: (id: number) => void;
}

export type ButtonsProps = {
    onButtonClicked: (button: string) => void;
    started: boolean;
};

export type DetailsProps = {
    activePlayer: string;
    turnTimer: number;
    value: string;
}

export type GameComponentState = {
    gameTimer: number;
    message: string;
    success: boolean;
    turnTimer: number;
    warning: boolean;
};

export type GameHistory = {
    duration: number;
    finishedTime: string;
    moves: number;
    winner: string;
};

export interface GameType extends BoxBase {
    id: number;
};

export type HistoryAction = {
    duration: number;
    finishedTime: Date;
}

export type IsPossibleToWin = {
    first: number;
    second: number
};

export type MatchedRows = {
    first: number;
    second: number;
};

export type MessageProps = {
    message: string;
    onRemoveAlert?: () => void;
    success: boolean;
    warning: boolean;
};

type Player = {
    active: boolean;
    name: string;
}

export type PlayerAction = {
    property: string;
    value: string;
}

export type PlayersProps = {
    onInputChanged: (property: string, name: string) => void;
    player1: string;
    player2: string;
}

export type ReducerAction = {
    payload: {
        boxes?: Box[];
        player?: PlayerAction;
        scoresList?: HistoryAction;
    }
    type: string;
};

export interface ScoresProps {
    scores: GameHistory[] | [];
}

export interface StoreState extends ScoresProps {
    boxes: Box[];
    finished: boolean;
    moves: number;
    next: GameType;
    player1: Player;
    player2: Player;
    started: boolean;
};
