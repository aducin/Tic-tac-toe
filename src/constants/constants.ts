import { GameType } from '../interfaces/interfaces';

export const DEFAULT_PLAYER1_NAME = 'Player 1';
export const DEFAULT_PLAYER2_NAME = 'Player 2';
export const EMPTY_HISTORY = 'Game history is empty at the moment';
export const FOOTER_TEXT = 'Tic-tac-toe footer';
export const GAME_NOT_STARTED = 'Click the button to start the game';
export const GAME_OVER = 'Game is over - no winner';
export const GAME_SUCCEDED = (winner: string, duration: number) =>
    `${winner} has won in ${duration.toFixed(2)} seconds!`;
export const HEADER_TEXT = 'Tic-tac-toe header';
export const MESSAGE_HEADER_SUCCESS = 'CONGRATULATIONS!';
export const MESSAGE_HEADER_WARNING = 'WARNING!';
export const MESSAGE_TIME = 3000;
export const MOBILE_MAX_WIDTH = '768px';
export const PLAYER_FIRST_TYPE = 'O';
export const PLAYER_SECOND_TYPE = 'X';
export const ROWS_TO_WIN = 3;
export const SCORES_LIST_HEADER = 'Score results';
export const TIME_IS_OVER = 'Time is over - next move';
export const TURN_TIME = 30;
export const WINNING_ROWS = [
    [0, 1, 2],
    [3, 4, 5],
    [6, 7, 8],
    [0, 3, 6],
    [1, 4, 7],
    [2, 5, 8],
    [0, 4, 8],
    [2, 4, 6]
];
export const GAMETYPES: GameType[] = [
    { 
        id: 1,
        icon: 'ImageO',
        iconWinning: 'ImageOWinning',
        value: PLAYER_FIRST_TYPE
    },
    {
        id: 2,
        icon: 'ImageX',
        iconWinning: 'ImageXWinning',
        value: PLAYER_SECOND_TYPE
    }
];
