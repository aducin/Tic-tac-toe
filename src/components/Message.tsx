import React from 'react';
import { MessageProps } from '../interfaces/interfaces';
import { MESSAGE_HEADER_SUCCESS, MESSAGE_HEADER_WARNING } from '../constants/constants';
import styles from '../assets/styles/message.module.scss';

const Message: React.FC<MessageProps> = (props: MessageProps) => {
    const containerClasses = [styles.container, 'alert'];
    const clearButton = props.onRemoveAlert ? (
        <button onClick={() => props.onRemoveAlert!()} type="button" className="close" aria-label="Close">
            <span aria-hidden="true">&times;</span>
        </button>
    ) : null;

    if (props.success) {
        containerClasses.push('alert-success');
    } else if (props.warning) {
        containerClasses.push('alert-danger');
    } else {
        containerClasses.push('alert-light');
    }
    const header = props.success ? MESSAGE_HEADER_SUCCESS : MESSAGE_HEADER_WARNING;
    return (
        <div className={containerClasses.join(' ')} role="alert">
            {clearButton}
            <h3 className={styles.header}>{header}</h3>
            <h4 className={styles.message}>{props.message}</h4>
        </div>
    );
}

export default React.memo(Message);
