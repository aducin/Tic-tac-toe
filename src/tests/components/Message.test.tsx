import React from "react";
import TestRenderer from 'react-test-renderer';
import { shallow } from "enzyme";
import { ShallowWrapper } from 'enzyme';
import Message from '../../components/Message';
import { MessageProps } from '../../types/types';
import { MESSAGE_HEADER_SUCCESS, MESSAGE_HEADER_WARNING } from '../../constants/constants';

const data: MessageProps = {
    message: 'You have won',
    onRemoveAlert: () => {},
    success: true,
    warning: false
}

const component = (
    <Message
        message={data.message}
        onRemoveAlert={data.onRemoveAlert}
        success={data.success}
        warning={data.warning}
    />
);
const testData = TestRenderer.create(component);

describe('Message component snapshot', () => {
    it('Message component renders correctly', async () => {
        expect(testData.toJSON()).toMatchSnapshot();
    });
});

describe('AccountDetails tests', () => {
    let wrapper: ShallowWrapper;
    beforeEach(() => {
        wrapper = shallow(component);
    });

    it('there is one h3 tag with a desired value', () => {
        const el = wrapper.find('h3');
        expect(el).toHaveLength(1);
        expect(el.text()).toEqual(MESSAGE_HEADER_SUCCESS);
    });

    it('there is one h4 tag with a desired value', () => {
        const el = wrapper.find('h4');
        expect(el).toHaveLength(1);
        expect(el.text()).toEqual(data.message);
    });

    it('there are alert classes attached to the div tag', () => {
        const el = wrapper.find('div');
        expect(el.hasClass('alert')).toBeTruthy;
        expect(el.hasClass('alert-danger')).toBeFalsy;
        expect(el.hasClass('alert-success')).toBeTruthy;
    });

    it('there should be a different header and alert class when success turned into warning', () => {
        wrapper.setProps({ success: false, warning: true });
        expect(wrapper.find('h3').text()).toEqual(MESSAGE_HEADER_WARNING);
        const el = wrapper.find('div');
        expect(el.hasClass('alert-danger')).toBeTruthy;
        expect(el.hasClass('alert-success')).toBeFalsy;
    }); 
});
