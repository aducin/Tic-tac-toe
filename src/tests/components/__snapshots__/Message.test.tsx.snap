// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message component snapshot Message component renders correctly 1`] = `
<div
  className="container alert alert-success"
  role="alert"
>
  <button
    aria-label="Close"
    className="close"
    onClick={[Function]}
    type="button"
  >
    <span
      aria-hidden="true"
    >
      Ã—
    </span>
  </button>
  <h3
    className="header"
  >
    CONGRATULATIONS!
  </h3>
  <h4
    className="message"
  >
    You have won
  </h4>
</div>
`;
